lib/charms/operator_libs_linux/v0/juju_systemd_notices.py

```diff
@@ -0,0 +1,485 @@
+    # It's basically urlencoded but instead of a %, it uses a _
+    path = name
+    for key, value in _DBUS_CHAR_MAPPINGS.items():
+        path = path.replace(value, key)
+
+    return f"/org/freedesktop/systemd1/unit/{path}"
+
+
+def _dbus_path_to_name(path: str) -> str:
+    """Convert the specified DBus path handle to a service name.
+
+    Args:
+        path: The DBus path to convert to service name.
+
+    Returns:
+        String containing the service name.
+    """
+    # DBUS Object names may only contain ASCII chars [A-Z][a-z][0-9]_
+    name = Path(path).name
+    for key, value in _DBUS_CHAR_MAPPINGS.items():
+        name = name.replace(key, value)
+
+    return name
+
+
+def _systemd_unit_changed(msg: Message) -> bool:
+    """Send Juju notification if systemd unit state changes on the DBus bus.
+
+    Invoked when a PropertiesChanged event occurs on an org.freedesktop.systemd1.Unit
+    object across the dbus. These events are sent whenever a unit changes state, including
+    starting and stopping.
+
+    Args:
+        msg: The message to process in the callback.
+
+    Returns:
+        True if the event is processed. False if otherwise.
+    """
+    _logger.debug(
+        f"Received message: path: {msg.path}, interface: {msg.interface}, member: {msg.member}"
+    )
+    service = _dbus_path_to_name(msg.path)
+    properties = msg.body[1]
+    if "ActiveState" not in properties:
+        return False
+
+    global _service_states
+    if service not in _service_states:
+        _logger.debug(f"Dropping event for unwatched service: {service}")
+        return False
+
+    curr_state = properties["ActiveState"].value
+    prev_state = _service_states[service]
+    # Drop transitioning and duplicate events
+    if curr_state.endswith("ing") or curr_state == prev_state:
+        _logger.debug(f"Dropping event - service: {service}, state: {curr_state}")
+        return False
+
+    _service_states[service] = curr_state
+    _logger.debug(f"Service {service} changed state to {curr_state}")
+    # Run the hook in a separate thread so the dbus notifications aren't
+    # blocked from being received.
+    asyncio.create_task(_send_juju_notification(service, curr_state))
+    return True
+
+
+async def _send_juju_notification(service: str, state: str) -> None:
+    """Invoke a Juju hook to notify an operator that a service state has changed.
+
+    Args:
+        service: The name of the service which has changed state.
+        state: The state of the service.
+    """
+    if service.endswith(".service"):
+        service = service[0 : -len(".service")]
+    if state == "active":
+        event_name = "started"
+    else:
+        event_name = "stopped"
+    hook = f"service-{service}-{event_name}"
+    cmd = ["/usr/bin/juju-exec", _juju_unit, f"hooks/{hook}"]
+
+    _logger.debug(f"Invoking hook {hook} with command: {' '.join(cmd)}")
+    process = await asyncio.create_subprocess_exec(
+        *cmd,
```

Comment:
Total nit, but this seems like it would be clearer on one line.

Reply:
I don't love this formatting either. This was a change that `black` made after I ran `tox run -e fmt`. I tried to change it back, but `black` really preferred this formatting for some reason. Let me see if there is something that I can do to make black NOT do this :smiley:

Reply:
Weird. I wonder if you had something else on the line before so it was wrapping? I just tried this on one line and Black doesn't seem to mind:

```
$ cat t.py
import asyncio


def foo(cmd):
    _logger.debug(f"Invoking hook {hook} with command: {' '.join(cmd)}")
    process = await asyncio.create_subprocess_exec(*cmd)
    await process.wait()
$ black --version
black, 23.7.0 (compiled: yes)
Python (CPython) 3.10.12
$ black t.py
All done! ✨ 🍰 ✨
1 file left unchanged.
```

Reply:
Just for the record, I don't think this is true:

> I believe this comma is here to prevent asyncio.create_subprocess_exec(...) from thinking the cmd list contains inputs for other values that asyncio.create_subprocess_exec(...) accepts

The comma is simply an optional trailing comma, and doesn't have any semantic meaning. See the `arguments:` rule in the [Python grammar](https://docs.python.org/3/reference/grammar.html), it has an optional `[',']` before the right parenthesis.

So `f(*cmd) is exactly equivalent to `f(*cmd,)`, the latter just has a trailing comma.

In addition, it wouldn't make sense for positional `*args` to "fill up" keyword args like `stdin=None` in this function's case.

You can show this by defining a function with the same args as `create_subprocess_exec` and then calling it with `*cmd`, and see that the other args aren't set:

```
>>> def f(program, *args, stdin=None, stdout=None, stderr=None, limit=None, **kwds):
...     print(program, args, stdin, stdout, stderr, limit, kwds)
...
>>> cmd = ['systemctl', 'start']
>>> f(*cmd)
systemctl ('start',) None None None None {}
>>> f(*cmd,)
systemctl ('start',) None None None None {}
```

Reply:
Changed in commit https://github.com/canonical/operator-libs-linux/pull/108/commits/3bd603a65b70ec45dfbc4ceff82114739179f9c6. After running the integration tests successfully I decided that the comma was no longer necessary.